### PR TITLE
AST, Parser and MongoOdm refactor

### LINKED ISSUE #10: sort(<+|-><property>) vs. sort(<property>,<direction>)
According to the [RQL specification](http://doc.apsstandard.org/2.1/spec/rql/), `sort()` requires a list of properties prepended with a `+` or `-`, not a property followed by a direction. The current parser does a great job of allowing this functionality,  no complaints there, but the `sort()` description in `QueryInterface.php` is misleading. Something akin to the following might be better:

```
/**
 * @param string $fields,... List of fields, each prefixed with <+|->
 *
 * @return mixed
 */
public function sort(/* ... */);
```

...and here is one possible implementation:

```
/**
 * {@inheritdoc}
 */
public function sort(/* ... */)
{

    $fields = func_get_args();
    $out = array();

    // Note that + must be URL encoded as %2B
    foreach( $fields as $field ) {
        switch( $field[0] ) {
            case '+' : array_push($out, substr($field, 1) . ' ASC' ); break;
            case '-' : array_push($out, substr($field, 1) . ' DESC' ); break;
            default: throw new \Exception("Must specify valid sort order for field {$field}"); break;
        }
    }

    $out = implode( ', ', $out );

    // And just for debugging...
    print_r( $out );
    exit;

}
```

Personally, I like implementing sorting as `sort($field, $direction = false)`, where `$direction` indicates whether the sorting should be reversed (i.e. `DESC`). This works great with the pipe/ajax functionality of the [SmartTable](http://lorenzofox3.github.io/smart-table-website/) module for AngularJS. (However, much like the suggested implementation described in `QueryInterface.php`, my implementation does not follow RQL spec.)

Here it is, if anyone out there is looking for more examples:

```
public function sort($field, $direction = false)
{
    // protected function wrapTable() turns objects.title into `objects`.title
    // It is meant for resolving ambiguity in joins by verifying against a list of allowed tables
    // e.g. foobar.title would not be allowed, b/c table `foobar` is not assoc. with the current entity
    $field = $this->wrapTable( $field );
    $direction = filter_var($direction, FILTER_VALIDATE_BOOLEAN);
    $direction = $direction ? 'DESC' : 'ASC';
    array_push( $this->order, "{$field} {$direction}" );
    // ...and then I retrieve the order with $queriable->get_order();
    // Of course, currently, I'm limited to sorting by one field at a time
}
```

I'm using [PHP ActiveRecord](http://www.phpactiverecord.org/) in my project. I've written a QueryInterface implementation for it, which generates an `$options['conditions']` array. I'd be happy to add it to the project as an example. I've never done a pull request before, but let me know if you are interested, and I'll figure out how to do it. Cheers!

### PR DESCRIPTION
**Pull request by hairmare**
The main focus on the is PR is the AST. The rest ist just here to show that it works ;) Have a look at the [branches README](https://github.com/libgraviton/php-rql-parser/blob/feature/ast-refactor/README.md).
- [x] refactor AST to more classes and interfaces
- [x] refactor Parser to use the Strategy pattern to do most of the work
- [x] make existing Visitor support the new AST design 
- [x] get #9 merged (it's also in here and I want this PR to be uncluttered, if we don't merge it i'll revert here as well)
- [x] add tests for example from #6 and make sure they work
- [x] refactor SortOperation (keeping #10 in mind, we support both cases in the AST)
- [x] get input on AST and maybe refactor it (i really like this verbose AST more than expected)

This fixes #10 and #6. It also makes us ready for nested queries and extending both the AST and the MongoDB visitors functionality. With this the AST is at a state that it can be used by others to implement their own visitors (ie. PHP_ActiveRecord or doctrine-orm).

With this and the work in #12 we will be ready to release `2.0.0-alpha2`.

For stabilising  `2.0.0` there is still work to be done:
- implement sugar syntax (no changes to the AST)
- support more things in the ast
- finish docs with topics on how to extend the ast and more
- split out mongo-odm visitor into its own repo (if we ever run into versioning problems with mongo this makes alot of sense, it also paves the way for integrating new visitors without cluttering our repo and issue tracker)
- also, more refactoring of the visitor to really make the best use of this ASTs interfaces
- expose proper error handling api and add optional logging

I will create the corresponding issues for these after this is merged.

### CODE REVIEW

File: src/AST/LimitOperation.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * @int
+     */
+    private $limit;
+
+    /**
+     * @int
+     */
+    private $skip;
+
```
Comment (lapistano) [inline src/AST/LimitOperation.php:24]: DocBlocks

File: src/AST/SortOperation.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * @license http://opensource.org/licenses/gpl-license.php GNU Public License
+ * @link    http://swisscom.ch
+ */
+class SortOperation extends AbstractOperation implements SortOperationInterface
+{
+    /**
+     * @var
+     */
+    private $fields = array();
+
```
Comment (lapistano) [inline src/AST/SortOperation.php:19]: docBlock

File: src/Parser/ParserUtil.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+use Graviton\Rql\Parser\ParserUtil;
+
+/**
+ * @author  List of contributors <https://github.com/libgraviton/php-rql-parser/graphs/contributors>
+ * @license http://opensource.org/licenses/gpl-license.php GNU Public License
+ * @link    http://swisscom.ch
+ */
+class ParserUtil
+{
+    public static function parseStart(Lexer &$lexer)
```
Comment (lapistano) [inline src/Parser/ParserUtil.php:25]: docBlocks

File: src/Parser/Strategy/ArrayOperationStrategy.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        $needsValue = true;
+        while ($this->hasValues()) {
+            if ($needsValue || ParserUtil::parseComma($this->lexer, true)) {
+                $operation->addValue(ParserUtil::getString($this->lexer, !$needsValue));
+                $needsValue = false;
+            }
+        }
+
```
Comment (lapistano) [inline src/Parser/Strategy/ArrayOperationStrategy.php:49]: DocBlocks

File: src/Parser/Strategy/LimitOperationStrategy.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        }
+        $operation->setLimit($fields[0]);
+        if (!empty($fields[1])) {
+            $operation->setSkip($fields[1]);
+        }
+        return $operation;
+    }
+
+    public function getAcceptedTypes()
```
Comment (lapistano) [inline src/Parser/Strategy/LimitOperationStrategy.php:52]: docblock

File: src/Parser/Strategy/ParsingStrategyInterface.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+use Graviton\Rql\AST\OperationInterface;
+
+/**
+ * @author  List of contributors <https://github.com/libgraviton/php-rql-parser/graphs/contributors>
+ * @license http://opensource.org/licenses/gpl-license.php GNU Public License
+ * @link    http://swisscom.ch
+ */
+interface ParsingStrategyInterface
+{
+    public function setParser(Parser &$parser);
```
Comment (lapistano) [inline src/Parser/Strategy/ParsingStrategyInterface.php:24]: docBlock

File: src/Parser/Strategy/PropertyOperationStrategy.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        ParserUtil::parseStart($this->lexer);
+        $operation->setProperty(ParserUtil::getString($this->lexer));
+        ParserUtil::parseComma($this->lexer);
+        $operation->setValue(ParserUtil::parseArgument($this->lexer));
+        ParserUtil::parseEnd($this->lexer);
+
+        return $operation;
+    }
+
+    public function getAcceptedTypes()
```
Comment (lapistano) [inline src/Parser/Strategy/PropertyOperationStrategy.php:43]: docBlocks

File: src/Parser/Strategy/QueryOperationStrategy.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            $hasQueries = $glimpse['type'] == Lexer::T_COMMA;
+            if (!$hasQueries) {
+                ParserUtil::parseEnd($this->lexer);
+            }
+        }
+
+        return $operation;
+    }
+
+    public function getAcceptedTypes()
```
Comment (lapistano) [inline src/Parser/Strategy/QueryOperationStrategy.php:54]: docBlock

File: src/Parser/Strategy/SortOperationStrategy.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public function getAcceptedTypes()
+    {
+        return array(
+            Lexer::T_SORT,
+        );
+    }
+
+    private function getTypeAndMove($token)
```
Comment (lapistano) [inline src/Parser/Strategy/SortOperationStrategy.php:64]: docBlocks

**Comment by hairmare** [comment]
@lapistano I just pushed the remaining docblocks. I ended up adding the ENTB coding standard to the checker since I needed a way to find them all. You'll probably want to merge this to #12 before we merge that to ensure that you don't have any ENTB violations in your code.

**Pull request merged**